File: api/src/main/java/com/wbrawner/budgetserver/category/CategoryController.java
package com.wbrawner.budgetserver.category;

import com.wbrawner.budgetserver.ErrorResponse;
import com.wbrawner.budgetserver.permission.Permission;
import com.wbrawner.budgetserver.permission.UserPermission;
import com.wbrawner.budgetserver.permission.UserPermissionRepository;
import com.wbrawner.budgetserver.transaction.TransactionRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.transaction.Transactional;
import java.util.List;
import java.util.stream.Collectors;

import static com.wbrawner.budgetserver.Utils.getCurrentUser;
import static com.wbrawner.budgetserver.Utils.getFirstOfMonth;

@RestController
@RequestMapping(path = "/categories")
@Transactional
class CategoryController {
    private final CategoryRepository categoryRepository;
    private final TransactionRepository transactionRepository;
    private final UserPermissionRepository userPermissionsRepository;

    CategoryController(CategoryRepository categoryRepository,
                       TransactionRepository transactionRepository,
                       UserPermissionRepository userPermissionsRepository) {
        this.categoryRepository = categoryRepository;
        this.transactionRepository = transactionRepository;
        this.userPermissionsRepository = userPermissionsRepository;
    }

    @GetMapping(path = "", produces = {MediaType.APPLICATION_JSON_VALUE})
    ResponseEntity<List<CategoryResponse>> getCategories(
            @RequestParam(name = "budgetIds", required = false) List<String> budgetIds,
            @RequestParam(name = "isExpense", required = false) Boolean isExpense,
            @RequestParam(name = "includeArchived", required = false) Boolean includeArchived,
            @RequestParam(name = "count", required = false) Integer count,
            @RequestParam(name = "page", required = false) Integer page,
            @RequestParam(name = "false", required = false) String sortBy,
            @RequestParam(name = "sortOrder", required = false) Sort.Direction sortOrder
    ) {
        List<UserPermission> userPermissions;
        if (budgetIds != null && !budgetIds.isEmpty()) {
            userPermissions = userPermissionsRepository.findAllByUserAndBudget_IdIn(
                    getCurrentUser(),
                    budgetIds,
                    PageRequest.of(page != null ? page : 0, count != null ? count : 1000)
            );
        } else {
            userPermissions = userPermissionsRepository.findAllByUser(getCurrentUser(), null);
        }
        var budgets = userPermissions.stream()
                .map(UserPermission::getBudget)
                .collect(Collectors.toList());

        var pageRequest = PageRequest.of(
                Math.min(0, page != null ? page - 1 : 0),
                count != null ? count : 1000,
                sortOrder != null ? sortOrder : Sort.Direction.ASC,
                sortBy != null ? sortBy : "title"
        );
        Boolean archived = includeArchived == null || includeArchived == false ? false : null;
        List<Category> categories = categoryRepository.findAllByBudgetIn(budgets, isExpense, archived, pageRequest);
        return ResponseEntity.ok(
                categories.stream()
                        .map(CategoryResponse::new)
                        .collect(Collectors.toList())
        );
    }

    @GetMapping(path = "/{id}", produces = {MediaType.APPLICATION_JSON_VALUE})
    ResponseEntity<CategoryResponse> getCategory(@PathVariable String id) {
        var budgets = userPermissionsRepository.findAllByUser(getCurrentUser(), null)
                .stream()
                .map(UserPermission::getBudget)
                .collect(Collectors.toList());
        var category = categoryRepository.findByBudgetInAndId(budgets, id).orElse(null);
        if (category == null) return ResponseEntity.notFound().build();
        return ResponseEntity.ok(new CategoryResponse(category));
    }

    @GetMapping(path = "/{id}/balance", produces = {MediaType.APPLICATION_JSON_VALUE})
    ResponseEntity<CategoryBalanceResponse> getCategoryBalance(@PathVariable String id) {
        var budgets = userPermissionsRepository.findAllByUser(getCurrentUser(), null)
                .stream()
                .map(UserPermission::getBudget)
                .collect(Collectors.toList());
        var category = categoryRepository.findByBudgetInAndId(budgets, id).orElse(null);
        if (category == null) {
            return ResponseEntity.notFound().build();
        }
        var sum = transactionRepository.sumBalanceByCategoryId(category.getId(), getFirstOfMonth());
        return ResponseEntity.ok(new CategoryBalanceResponse(category.getId(), sum));
    }

    @PostMapping(path = "", consumes = {MediaType.APPLICATION_JSON_VALUE}, produces = {MediaType.APPLICATION_JSON_VALUE})
    ResponseEntity<Object> newCategory(@RequestBody NewCategoryRequest request) {
        var userResponse = userPermissionsRepository.findByUserAndBudget_Id(getCurrentUser(), request.getBudgetId())
                .orElse(null);
        if (userResponse == null) {
            return ResponseEntity.badRequest().body(new ErrorResponse("Invalid budget ID"));
        }
        if (userResponse.getPermission().isNotAtLeast(Permission.WRITE)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
        }
        var budget = userResponse.getBudget();
        return ResponseEntity.ok(new CategoryResponse(categoryRepository.save(new Category(
                request.getTitle(),
                request.getDescription(),
                request.getAmount(),
                budget,
                request.getExpense()
        ))));
    }

    @PutMapping(path = "/{id}", consumes = {MediaType.APPLICATION_JSON_VALUE}, produces = {MediaType.APPLICATION_JSON_VALUE})
    ResponseEntity<CategoryResponse> updateCategory(@PathVariable String id, @RequestBody UpdateCategoryRequest request) {
        var category = categoryRepository.findById(id).orElse(null);
        if (category == null) return ResponseEntity.notFound().build();
        var userPermission = userPermissionsRepository.findByUserAndBudget_Id(getCurrentUser(), category.getBudget().getId()).orElse(null);
        if (userPermission == null) return ResponseEntity.notFound().build();
        if (userPermission.getPermission().isNotAtLeast(Permission.WRITE)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
        }
        if (request.getTitle() != null) {
            category.setTitle(request.getTitle());
        }
        if (request.getDescription() != null) {
            category.setDescription(request.getDescription());
        }
        if (request.getAmount() != null) {
            category.setAmount(request.getAmount());
        }
        if (request.getExpense() != null) {
            category.setExpense(request.getExpense());
        }
        if (request.getArchived() != null) {
            category.setArchived(request.getArchived());
        }
        return ResponseEntity.ok(new CategoryResponse(categoryRepository.save(category)));
    }

    @DeleteMapping(path = "/{id}", produces = {MediaType.TEXT_PLAIN_VALUE})
    ResponseEntity<Void> deleteCategory(@PathVariable String id) {
        var category = categoryRepository.findById(id).orElse(null);
        if (category == null) return ResponseEntity.notFound().build();
        var userPermission = userPermissionsRepository.findByUserAndBudget_Id(getCurrentUser(), category.getBudget().getId()).orElse(null);
        if (userPermission == null) return ResponseEntity.notFound().build();
        if (userPermission.getPermission().isNotAtLeast(Permission.WRITE)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
        }
        transactionRepository.findAllByBudgetAndCategory(userPermission.getBudget(), category)
                .forEach(transaction -> {
                    transaction.setCategory(null);
                    transactionRepository.save(transaction);
                });
        categoryRepository.delete(category);
        return ResponseEntity.ok().build();
    }
}

File: app/src/main/kotlin/com/wbrawner/twigs/server/category/CategoryController.kt
package com.wbrawner.twigs.server.category

import com.wbrawner.twigs.server.ErrorResponse
import com.wbrawner.twigs.server.currentUser
import com.wbrawner.twigs.server.firstOfMonth
import com.wbrawner.twigs.server.permission.Permission
import com.wbrawner.twigs.server.permission.UserPermission
import com.wbrawner.twigs.server.permission.UserPermissionRepository
import com.wbrawner.twigs.server.transaction.Transaction
import com.wbrawner.twigs.server.transaction.TransactionRepository
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.function.Consumer
import java.util.stream.Collectors
import javax.transaction.Transactional

@RestController
@RequestMapping(path = ["/categories"])
@Transactional
open class CategoryController(
    private val categoryRepository: CategoryRepository,
    private val transactionRepository: TransactionRepository,
    private val userPermissionsRepository: UserPermissionRepository
) {
    @GetMapping(path = [""], produces = [MediaType.APPLICATION_JSON_VALUE])
    open fun getCategories(
        @RequestParam(name = "budgetIds", required = false) budgetIds: List<String?>?,
        @RequestParam(name = "isExpense", required = false) isExpense: Boolean?,
        @RequestParam(name = "includeArchived", required = false) includeArchived: Boolean?,
        @RequestParam(name = "count", required = false) count: Int?,
        @RequestParam(name = "page", required = false) page: Int?,
        @RequestParam(name = "false", required = false) sortBy: String?,
        @RequestParam(name = "sortOrder", required = false) sortOrder: Sort.Direction?
    ): ResponseEntity<List<CategoryResponse>> {
        val userPermissions: List<UserPermission>
        userPermissions = if (budgetIds != null && !budgetIds.isEmpty()) {
            userPermissionsRepository.findAllByUserAndBudget_IdIn(
                currentUser,
                budgetIds,
                PageRequest.of(page ?: 0, count ?: 1000)
            )
        } else {
            userPermissionsRepository.findAllByUser(currentUser, null)
        }
        val budgets = userPermissions.stream()
            .map { obj: UserPermission -> obj.budget }
            .collect(Collectors.toList())
        val pageRequest = PageRequest.of(
            Math.min(0, if (page != null) page - 1 else 0),
            count ?: 1000,
            sortOrder ?: Sort.Direction.ASC,
            sortBy ?: "title"
        )
        val archived = if (includeArchived == null || includeArchived == false) false else null
        val categories = categoryRepository.findAllByBudgetIn(budgets, isExpense, archived, pageRequest)
        return ResponseEntity.ok(
            categories.stream()
                .map { category: Category -> CategoryResponse(category) }
                .collect(Collectors.toList())
        )
    }

    @GetMapping(path = ["/{id}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    open fun getCategory(@PathVariable id: String?): ResponseEntity<CategoryResponse> {
        val budgets = userPermissionsRepository.findAllByUser(currentUser, null)
            .stream()
            .map { obj: UserPermission -> obj.budget }
            .collect(Collectors.toList())
        val category = categoryRepository.findByBudgetInAndId(budgets, id).orElse(null)
            ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(CategoryResponse(category))
    }

    @GetMapping(path = ["/{id}/balance"], produces = [MediaType.APPLICATION_JSON_VALUE])
    open fun getCategoryBalance(@PathVariable id: String?): ResponseEntity<CategoryBalanceResponse> {
        val budgets = userPermissionsRepository.findAllByUser(currentUser, null)
            .stream()
            .map { obj: UserPermission -> obj.budget }
            .collect(Collectors.toList())
        val category = categoryRepository.findByBudgetInAndId(budgets, id).orElse(null)
            ?: return ResponseEntity.notFound().build()
        val sum = transactionRepository.sumBalanceByCategoryId(category.id, firstOfMonth)
        return ResponseEntity.ok(CategoryBalanceResponse(category.id, sum))
    }

    @PostMapping(
        path = [""],
        consumes = [MediaType.APPLICATION_JSON_VALUE],
        produces = [MediaType.APPLICATION_JSON_VALUE]
    )
    open fun newCategory(@RequestBody request: NewCategoryRequest): ResponseEntity<Any> {
        val userResponse = userPermissionsRepository.findByUserAndBudget_Id(currentUser, request.budgetId)
            .orElse(null) ?: return ResponseEntity.badRequest().body(ErrorResponse("Invalid budget ID"))
        if (userResponse.permission.isNotAtLeast(Permission.WRITE)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build()
        }
        val budget = userResponse.budget
        return ResponseEntity.ok(
            CategoryResponse(
                categoryRepository.save(
                    Category(
                        title = request.title,
                        description = request.description,
                        amount = request.amount,
                        budget = budget,
                        expense = request.expense
                    )
                )
            )
        )
    }

    @PutMapping(
        path = ["/{id}"],
        consumes = [MediaType.APPLICATION_JSON_VALUE],
        produces = [MediaType.APPLICATION_JSON_VALUE]
    )
    open fun updateCategory(
        @PathVariable id: String,
        @RequestBody request: UpdateCategoryRequest
    ): ResponseEntity<CategoryResponse> {
        val category = categoryRepository.findById(id).orElse(null)
            ?: return ResponseEntity.notFound().build()
        val userPermission = userPermissionsRepository.findByUserAndBudget_Id(
            currentUser,
            category.budget!!.id
        ).orElse(null)
            ?: return ResponseEntity.notFound().build()
        if (userPermission.permission.isNotAtLeast(Permission.WRITE)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build()
        }
        if (request.title != null) {
            category.title = request.title
        }
        if (request.description != null) {
            category.description = request.description
        }
        if (request.amount != null) {
            category.amount = request.amount
        }
        if (request.expense != null) {
            category.expense = request.expense
        }
        if (request.archived != null) {
            category.archived = request.archived
        }
        return ResponseEntity.ok(CategoryResponse(categoryRepository.save(category)))
    }

    @DeleteMapping(path = ["/{id}"], produces = [MediaType.TEXT_PLAIN_VALUE])
    open fun deleteCategory(@PathVariable id: String): ResponseEntity<Void> {
        val category = categoryRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        val userPermission =
            userPermissionsRepository.findByUserAndBudget_Id(currentUser, category.budget!!.id).orElse(null)
                ?: return ResponseEntity.notFound().build()
        if (userPermission.permission.isNotAtLeast(Permission.WRITE)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build()
        }
        transactionRepository.findAllByBudgetAndCategory(userPermission.budget, category)
            .forEach(Consumer { transaction: Transaction ->
                transaction.category = null
                transactionRepository.save(transaction)
            })
        categoryRepository.delete(category)
        return ResponseEntity.ok().build()
    }
}